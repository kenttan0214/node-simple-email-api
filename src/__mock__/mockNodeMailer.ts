export const mockCreateTestAccount = jest.fn(() => ({ user: 'test-account', pass: 'test-password' }));
export const mockSendMail = jest.fn();
export const mockCreateTransport = jest.fn(() => ({ sendMail: mockSendMail }));
export const mockGetTestMessageUrl = jest.fn(() => 'https://unit-test-mock-email-url.example.com');

jest.mock('nodemailer', () => ({
  createTestAccount: mockCreateTestAccount,
  createTransport: mockCreateTransport,
  getTestMessageUrl: mockGetTestMessageUrl,
}));
