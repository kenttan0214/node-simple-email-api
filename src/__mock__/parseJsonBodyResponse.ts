// a simple util function to parse the unit test JSON response into Object
export const parseJsonBody = (response: { body: string }): Record<string, any> => {
  return {
    ...response,
    body: JSON.parse(response.body),
  };
};
