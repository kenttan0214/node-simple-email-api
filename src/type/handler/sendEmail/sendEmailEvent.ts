import { APIGatewayProxyEvent } from 'aws-lambda';

export type EmailBody = {
  subject: string;
  content: string;
  recipients: string[];
};

export type SendEmailEvent = {
  body: EmailBody;
} & APIGatewayProxyEvent;
