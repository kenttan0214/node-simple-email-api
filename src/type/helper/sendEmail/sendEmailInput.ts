export type EmailInput = {
  subject: string;
  content: string;
  recipients: string[];
};
