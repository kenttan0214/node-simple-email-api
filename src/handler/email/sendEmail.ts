import { APIGatewayProxyResult } from 'aws-lambda';
import middleware from '../../middleware';
import { SendEmailEvent } from '../../type/handler/sendEmail/sendEmailEvent';
import { badRequest, ok } from '../../util/apiGatewayResponse';
import sendEmailHelper from '../../helper/email/sendEmail';

export const inputSchema = {
  type: 'object',
  required: ['body'],
  properties: {
    body: {
      type: 'object',
      properties: {
        subject: {
          type: 'string',
        },
        content: {
          type: 'string',
        },
        recipients: {
          type: 'array',
          minItems: 1,
          uniqueItems: true,
          items: {
            type: 'string',
            format: 'email',
          },
        },
      },
      required: ['subject', 'content', 'recipients'],
    },
  },
};

const sendEmail = async (event: SendEmailEvent): Promise<APIGatewayProxyResult> => {
  try {
    const {
      body: { subject, content, recipients },
    } = event;

    const response = await sendEmailHelper({ subject, content, recipients });

    return ok(response);
  } catch (error) {
    return badRequest(error);
  }
};

export const handler = middleware(sendEmail, inputSchema);
