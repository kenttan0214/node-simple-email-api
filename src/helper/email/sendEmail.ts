import { createTestAccount, getTestMessageUrl } from 'nodemailer';
import { EmailInput } from '../../type/helper/sendEmail/sendEmailInput';
import smtpTransporter from '../../util/smtpTransporter';

export default async ({ subject, content, recipients }: EmailInput): Promise<{ preview: string | false }> => {
  // use nodemailer test account, actual use case will create own mail acccount such as Gmail
  const testAccount = await createTestAccount();

  // create reusable transporter object using the default SMTP transport
  const transporter = smtpTransporter(testAccount);

  // send mail with defined transport object
  const info = await transporter.sendMail({
    from: '"Fred Foo" <foo@example.com>', // sender address
    to: recipients.toString(), // list of recipients
    subject, // Subject line
    text: content, // plain text body
  });

  return {
    // Preview only available when sending through an Ethereal account
    preview: getTestMessageUrl(info),
  };
};
