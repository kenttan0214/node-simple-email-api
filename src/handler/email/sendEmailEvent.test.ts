import { parseJsonBody } from '../../__mock__/parseJsonBodyResponse';
import '../../__mock__/mockNodeMailer';
import * as sendEmailHelper from '../../helper/email/sendEmail';
import { handler } from './sendEmail';

describe('sendEmail hanlder', () => {
  it('should return 200 response with email preview URL', async () => {
    const event = {
      body: {
        subject: 'foo subject',
        content: 'foo content',
        recipients: ['foo@example.com', 'bar@example.com'],
      },
    };

    const response = await handler(event);
    expect(parseJsonBody(response)).toMatchSnapshot();
  });

  it('should return 400 error when there is any unknown exception thrown in helper fuction', async () => {
    const spy = jest.spyOn(sendEmailHelper, 'default');
    // Simulate situation where helper throws unknown, unexpected errors
    spy.mockRejectedValueOnce('unknown error');

    const event = {
      body: {
        subject: 'foo subject',
        content: 'foo content',
        recipients: ['foo@example.com'],
      },
    };

    const response = await handler(event);
    expect(parseJsonBody(response)).toMatchSnapshot();
  });
});

describe('input validation', () => {
  it('should return 400 error when subject is not provided', async () => {
    const event = {
      body: {
        content: 'foo content',
        recipients: ['foo@example.com', 'bar@example.com'],
      },
    };

    const response = await handler(event);
    expect(parseJsonBody(response)).toMatchSnapshot();
  });

  it('should return 400 error when content is not provided', async () => {
    const event = {
      body: {
        subject: 'foo subject',
        recipients: ['foo@example.com', 'bar@example.com'],
      },
    };

    const response = await handler(event);
    expect(parseJsonBody(response)).toMatchSnapshot();
  });

  it('should return 400 error when recipients is not provided', async () => {
    const event = {
      body: {
        subject: 'foo subject',
        content: 'foo content',
      },
    };

    const response = await handler(event);
    expect(parseJsonBody(response)).toMatchSnapshot();
  });

  it('should return 400 error when recipients is an empty array', async () => {
    const event = {
      body: {
        subject: 'foo subject',
        content: 'foo content',
        recipients: [],
      },
    };

    const response = await handler(event);
    expect(parseJsonBody(response)).toMatchSnapshot();
  });

  it('should return 400 error when recipients have duplicate entry', async () => {
    const event = {
      body: {
        subject: 'foo subject',
        content: 'foo content',
        recipients: ['foo@example.com', 'foo@example.com'],
      },
    };

    const response = await handler(event);
    expect(parseJsonBody(response)).toMatchSnapshot();
  });
});
