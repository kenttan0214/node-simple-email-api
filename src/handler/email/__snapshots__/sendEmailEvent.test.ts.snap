// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input validation should return 400 error when content is not provided 1`] = `
Object {
  "body": Object {
    "details": Array [
      Object {
        "instancePath": "/body",
        "keyword": "required",
        "message": "must have required property content",
        "params": Object {
          "missingProperty": "content",
        },
        "schemaPath": "#/properties/body/required",
      },
    ],
    "message": "Event object failed validation",
  },
  "statusCode": 400,
}
`;

exports[`input validation should return 400 error when recipients have duplicate entry 1`] = `
Object {
  "body": Object {
    "details": Array [
      Object {
        "instancePath": "/body/recipients",
        "keyword": "uniqueItems",
        "message": "must NOT have duplicate items (items ## 1 and 0 are identical)",
        "params": Object {
          "i": 0,
          "j": 1,
        },
        "schemaPath": "#/properties/body/properties/recipients/uniqueItems",
      },
    ],
    "message": "Event object failed validation",
  },
  "statusCode": 400,
}
`;

exports[`input validation should return 400 error when recipients is an empty array 1`] = `
Object {
  "body": Object {
    "details": Array [
      Object {
        "instancePath": "/body/recipients",
        "keyword": "minItems",
        "message": "must NOT have less than 1 item",
        "params": Object {
          "limit": 1,
        },
        "schemaPath": "#/properties/body/properties/recipients/minItems",
      },
    ],
    "message": "Event object failed validation",
  },
  "statusCode": 400,
}
`;

exports[`input validation should return 400 error when recipients is not provided 1`] = `
Object {
  "body": Object {
    "details": Array [
      Object {
        "instancePath": "/body",
        "keyword": "required",
        "message": "must have required property recipients",
        "params": Object {
          "missingProperty": "recipients",
        },
        "schemaPath": "#/properties/body/required",
      },
    ],
    "message": "Event object failed validation",
  },
  "statusCode": 400,
}
`;

exports[`input validation should return 400 error when subject is not provided 1`] = `
Object {
  "body": Object {
    "details": Array [
      Object {
        "instancePath": "/body",
        "keyword": "required",
        "message": "must have required property subject",
        "params": Object {
          "missingProperty": "subject",
        },
        "schemaPath": "#/properties/body/required",
      },
    ],
    "message": "Event object failed validation",
  },
  "statusCode": 400,
}
`;

exports[`sendEmail hanlder should return 200 response with email preview URL 1`] = `
Object {
  "body": Object {
    "preview": "https://unit-test-mock-email-url.example.com",
  },
  "headers": Object {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 200,
}
`;

exports[`sendEmail hanlder should return 400 error when there is any unknown exception thrown in helper fuction 1`] = `
Object {
  "body": Object {
    "errorMessage": "Something Went Wrong",
  },
  "headers": Object {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 400,
}
`;
